File: android/src/main/kotlin/com/ezoic/ezoic_flutter_sdk/EzoicNativeAdPlatformView.kt
package com.ezoic.ezoic_flutter_sdk

import android.content.Context
import android.graphics.Typeface
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.ezoic.ads.sdk.adunits.EzoicNativeAd
import com.ezoic.ads.sdk.adunits.EzoicNativeAdListener
import com.ezoic.ads.sdk.adunits.EzoicNativeAdLoadListener
import com.ezoic.ads.sdk.core.EzoicError
import com.google.android.gms.ads.nativead.MediaView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class EzoicNativeAdViewFactory(private val messenger: BinaryMessenger) :
  PlatformViewFactory(StandardMessageCodec.INSTANCE) {
  override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
    @Suppress("UNCHECKED_CAST")
    val params = (args as? Map<String, Any?>) ?: emptyMap()
    return EzoicNativeAdPlatformView(context, viewId, params, messenger)
  }
}

class EzoicNativeAdPlatformView(
  private val context: Context,
  viewId: Int,
  params: Map<String, Any?>,
  messenger: BinaryMessenger,
) : PlatformView {

  private val container = FrameLayout(context)
  private val channel = MethodChannel(messenger, "com.ezoic/ezoic_native_ad_view_$viewId")
  private val adUnitId = (params["adUnitIdentifier"] as? String)?.toIntOrNull() ?: 0
  private var ezoicNativeAd: EzoicNativeAd? = null
  private var loadStarted = false
  private var disposed = false

  init {
    // The load is triggered by an explicit "load" call from Dart rather than
    // here, so the Dart-side handler is attached before the native SDK can
    // fail synchronously (e.g. EzoicError.NotInitialized) and drop onError.
    channel.setMethodCallHandler { call, result ->
      when (call.method) {
        "load" -> {
          startLoad()
          result.success(null)
        }
        else -> result.notImplemented()
      }
    }
  }

  private fun startLoad() {
    // A second "load" is a no-op success; the guard survives across calls.
    if (loadStarted) return
    loadStarted = true
    EzoicNativeAd.load(context, adUnitId, object : EzoicNativeAdLoadListener {
      override fun onNativeAdLoaded(nativeAd: EzoicNativeAd) {
        // dispose() may run mid-load; the SDK callback and dispose both arrive
        // on the main thread, so this check is race-free. Destroy the ad the
        // SDK just handed us to avoid leaking it (there is no deinit).
        if (disposed) {
          nativeAd.destroy()
          return
        }
        ezoicNativeAd = nativeAd
        val gmaAd = nativeAd.nativeAd ?: run {
          channel.invokeMethod(
            "onError", mapOf("message" to "Native ad loaded without content", "code" to 0))
          return
        }
        // Attach the lifecycle listener before the rendered NativeAdView
        // registers — the impression fires as soon as the view is displayed.
        nativeAd.listener = object : EzoicNativeAdListener {
          override fun onNativeAdImpression(nativeAd: EzoicNativeAd) {
            channel.invokeMethod("onImpression", null)
          }
          override fun onNativeAdClicked(nativeAd: EzoicNativeAd) {
            channel.invokeMethod("onClick", null)
          }
          override fun onNativeAdOpened(nativeAd: EzoicNativeAd) {
            channel.invokeMethod("onOpen", null)
          }
          override fun onNativeAdClosed(nativeAd: EzoicNativeAd) {
            channel.invokeMethod("onClose", null)
          }
        }
        val adView = buildTemplate(gmaAd)
        container.removeAllViews()
        container.addView(adView)
        channel.invokeMethod("onLoad", null)
      }

      override fun onNativeAdFailedToLoad(error: EzoicError) {
        if (disposed) return
        channel.invokeMethod("onError", mapOf("message" to error.message, "code" to error.code))
      }
    })
  }

  /**
   * Builds a template [NativeAdView] entirely in code (the plugin ships no
   * `res/` layouts). Layout: a vertical column of a header row (icon +
   * headline/advertiser), a [MediaView], the body text and a call-to-action
   * button. Only the asset views actually present on [gmaAd] are created and
   * registered; [NativeAdView.setNativeAd] is called last, as GMA requires.
   */
  private fun buildTemplate(gmaAd: NativeAd): NativeAdView {
    val adView = NativeAdView(context)

    val root = LinearLayout(context).apply {
      orientation = LinearLayout.VERTICAL
      layoutParams = FrameLayout.LayoutParams(
        FrameLayout.LayoutParams.MATCH_PARENT,
        FrameLayout.LayoutParams.WRAP_CONTENT,
      )
      val pad = dp(8)
      setPadding(pad, pad, pad, pad)
    }

    val headerRow = LinearLayout(context).apply {
      orientation = LinearLayout.HORIZONTAL
      layoutParams = LinearLayout.LayoutParams(
        LinearLayout.LayoutParams.MATCH_PARENT,
        LinearLayout.LayoutParams.WRAP_CONTENT,
      )
    }

    var iconView: ImageView? = null
    gmaAd.icon?.drawable?.let { drawable ->
      val iv = ImageView(context).apply {
        layoutParams = LinearLayout.LayoutParams(dp(40), dp(40))
        setImageDrawable(drawable)
      }
      headerRow.addView(iv)
      iconView = iv
    }

    val textColumn = LinearLayout(context).apply {
      orientation = LinearLayout.VERTICAL
      layoutParams = LinearLayout.LayoutParams(
        0,
        LinearLayout.LayoutParams.WRAP_CONTENT,
        1f,
      ).apply { leftMargin = dp(8) }
    }

    var headlineView: TextView? = null
    gmaAd.headline?.let { text ->
      val tv = TextView(context).apply {
        this.text = text
        setTypeface(typeface, Typeface.BOLD)
        textSize = 16f
      }
      textColumn.addView(tv)
      headlineView = tv
    }

    var advertiserView: TextView? = null
    gmaAd.advertiser?.let { text ->
      val tv = TextView(context).apply {
        this.text = text
        textSize = 12f
      }
      textColumn.addView(tv)
      advertiserView = tv
    }

    headerRow.addView(textColumn)
    root.addView(headerRow)

    var mediaView: MediaView? = null
    gmaAd.mediaContent?.let { content ->
      val mv = MediaView(context).apply {
        layoutParams = LinearLayout.LayoutParams(
          LinearLayout.LayoutParams.MATCH_PARENT,
          dp(175),
        ).apply { topMargin = dp(8) }
        mediaContent = content
      }
      root.addView(mv)
      mediaView = mv
    }

    var bodyView: TextView? = null
    gmaAd.body?.let { text ->
      val tv = TextView(context).apply {
        this.text = text
        textSize = 14f
        setPadding(0, dp(8), 0, 0)
      }
      root.addView(tv)
      bodyView = tv
    }

    var callToActionView: Button? = null
    gmaAd.callToAction?.let { text ->
      val btn = Button(context).apply {
        this.text = text
        layoutParams = LinearLayout.LayoutParams(
          LinearLayout.LayoutParams.MATCH_PARENT,
          LinearLayout.LayoutParams.WRAP_CONTENT,
        ).apply { topMargin = dp(8) }
      }
      root.addView(btn)
      callToActionView = btn
    }

    adView.addView(root)
    // Register only the asset views that were populated; a null assignment
    // leaves that asset unregistered.
    adView.headlineView = headlineView
    adView.bodyView = bodyView
    adView.iconView = iconView
    adView.advertiserView = advertiserView
    adView.callToActionView = callToActionView
    adView.mediaView = mediaView
    adView.setNativeAd(gmaAd)
    return adView
  }

  private fun dp(value: Int): Int =
    (value * context.resources.displayMetrics.density).toInt()

  override fun getView(): View = container

  override fun dispose() {
    disposed = true
    // Clear the inbound handler: the messenger keys handlers by channel name
    // for the engine's lifetime and the closure retains this view (and its
    // Context) — view ids are never reused, so an uncleared handler leaks.
    channel.setMethodCallHandler(null)
    ezoicNativeAd?.destroy()
    ezoicNativeAd = null
    container.removeAllViews()
  }
}

File: ios/Classes/EzoicNativeAdPlatformView.swift
import Flutter
import UIKit
import EzoicAdsSDKBinary
import GoogleMobileAds

class EzoicNativeAdViewFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger
  init(messenger: FlutterBinaryMessenger) { self.messenger = messenger }

  func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
    return EzoicNativeAdPlatformView(frame: frame, viewId: viewId, args: args, messenger: messenger)
  }

  // Required or creationParams arrive nil on the native side.
  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol { FlutterStandardMessageCodec.sharedInstance() }
}

class EzoicNativeAdPlatformView: NSObject, FlutterPlatformView, EzoicNativeAdDelegate {
  private let container = UIView()
  private let channel: FlutterMethodChannel
  private let adUnitId: Int
  private var ezoicNativeAd: EzoicNativeAd?
  private var adView: NativeAdView?
  private var loadStarted = false

  init(frame: CGRect, viewId: Int64, args: Any?, messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: "com.ezoic/ezoic_native_ad_view_\(viewId)", binaryMessenger: messenger)
    let params = args as? [String: Any]
    adUnitId = Int(params?["adUnitIdentifier"] as? String ?? "") ?? 0
    super.init()
    // The load is triggered by an explicit "load" call from Dart rather than
    // here, so the Dart-side handler is attached before the native SDK can
    // fail synchronously (e.g. not initialized) and drop onError.
    channel.setMethodCallHandler { [weak self] call, result in
      guard let self = self else {
        result(nil)
        return
      }
      switch call.method {
      case "load":
        self.startLoad()
        result(nil)
      default:
        result(FlutterMethodNotImplemented)
      }
    }
  }

  private func startLoad() {
    // A second "load" is a no-op success; the guard survives across calls.
    if loadStarted { return }
    loadStarted = true
    EzoicNativeAd.load(adUnitIdentifier: adUnitId) { [weak self] result in
      guard let self = self else { return }
      switch result {
      case .success(let ad):
        self.ezoicNativeAd = ad
        guard let gmaAd = ad.nativeAd else {
          self.channel.invokeMethod(
            "onError", arguments: ["message": "Native ad loaded without content", "code": 0])
          return
        }
        // Attach the delegate before rendering so the impression, which fires
        // as soon as the NativeAdView is displayed, is delivered.
        ad.delegate = self
        self.render(gmaAd)
        self.channel.invokeMethod("onLoad", arguments: nil)
      case .failure(let error):
        self.channel.invokeMethod(
          "onError", arguments: ["message": error.localizedDescription, "code": error.code])
      }
    }
  }

  func view() -> UIView { container }

  /// Builds a template `NativeAdView` in code (mirrors the Android template):
  /// a header row (icon + headline/advertiser), a `MediaView`, the body text
  /// and a call-to-action button. The optional text/image assets are only
  /// created and registered when present, but the `MediaView` is always built:
  /// on GMA 12 `NativeAd.mediaContent` is non-optional and the media view is a
  /// required asset (unlike Android, where `mediaContent` is `@Nullable`).
  /// `adView.nativeAd` is assigned last.
  private func render(_ gmaAd: GoogleMobileAds.NativeAd) {
    let adView = NativeAdView()
    adView.translatesAutoresizingMaskIntoConstraints = false

    let mainStack = UIStackView()
    mainStack.axis = .vertical
    mainStack.spacing = 8
    mainStack.translatesAutoresizingMaskIntoConstraints = false

    let headerRow = UIStackView()
    headerRow.axis = .horizontal
    headerRow.spacing = 8
    headerRow.alignment = .center

    if let image = gmaAd.icon?.image {
      let iconView = UIImageView(image: image)
      iconView.translatesAutoresizingMaskIntoConstraints = false
      NSLayoutConstraint.activate([
        iconView.widthAnchor.constraint(equalToConstant: 40),
        iconView.heightAnchor.constraint(equalToConstant: 40),
      ])
      headerRow.addArrangedSubview(iconView)
      adView.iconView = iconView
    }

    let textColumn = UIStackView()
    textColumn.axis = .vertical

    if let headline = gmaAd.headline {
      let label = UILabel()
      label.text = headline
      label.font = .boldSystemFont(ofSize: 16)
      label.numberOfLines = 0
      textColumn.addArrangedSubview(label)
      adView.headlineView = label
    }

    if let advertiser = gmaAd.advertiser {
      let label = UILabel()
      label.text = advertiser
      label.font = .systemFont(ofSize: 12)
      textColumn.addArrangedSubview(label)
      adView.advertiserView = label
    }

    headerRow.addArrangedSubview(textColumn)
    mainStack.addArrangedSubview(headerRow)

    let mediaView = MediaView()
    mediaView.mediaContent = gmaAd.mediaContent
    mediaView.translatesAutoresizingMaskIntoConstraints = false
    // Priority 999 so a caller-supplied SizedBox shorter than the template's
    // natural height breaks this constraint instead of spamming
    // unsatisfiable-constraint logs (Android clips silently in that case).
    let mediaHeight = mediaView.heightAnchor.constraint(equalToConstant: 175)
    mediaHeight.priority = UILayoutPriority(999)
    mediaHeight.isActive = true
    mainStack.addArrangedSubview(mediaView)
    adView.mediaView = mediaView

    if let body = gmaAd.body {
      let label = UILabel()
      label.text = body
      label.font = .systemFont(ofSize: 14)
      label.numberOfLines = 0
      mainStack.addArrangedSubview(label)
      adView.bodyView = label
    }

    if let cta = gmaAd.callToAction {
      let button = UIButton(type: .system)
      button.setTitle(cta, for: .normal)
      // The NativeAdView handles the tap; the button must not intercept it.
      button.isUserInteractionEnabled = false
      mainStack.addArrangedSubview(button)
      adView.callToActionView = button
    }

    adView.addSubview(mainStack)
    NSLayoutConstraint.activate([
      mainStack.topAnchor.constraint(equalTo: adView.topAnchor, constant: 8),
      mainStack.leadingAnchor.constraint(equalTo: adView.leadingAnchor, constant: 8),
      mainStack.trailingAnchor.constraint(equalTo: adView.trailingAnchor, constant: -8),
      mainStack.bottomAnchor.constraint(equalTo: adView.bottomAnchor, constant: -8),
    ])

    container.addSubview(adView)
    NSLayoutConstraint.activate([
      adView.topAnchor.constraint(equalTo: container.topAnchor),
      adView.leadingAnchor.constraint(equalTo: container.leadingAnchor),
      adView.trailingAnchor.constraint(equalTo: container.trailingAnchor),
      adView.bottomAnchor.constraint(equalTo: container.bottomAnchor),
    ])

    adView.nativeAd = gmaAd
    self.adView = adView
  }

  // MARK: - EzoicNativeAdDelegate

  func nativeAdDidRecordImpression(_ nativeAd: EzoicNativeAd) {
    channel.invokeMethod("onImpression", arguments: nil)
  }

  func nativeAdDidRecordClick(_ nativeAd: EzoicNativeAd) {
    channel.invokeMethod("onClick", arguments: nil)
  }

  func nativeAdWillPresentScreen(_ nativeAd: EzoicNativeAd) {
    channel.invokeMethod("onOpen", arguments: nil)
  }

  func nativeAdDidDismissScreen(_ nativeAd: EzoicNativeAd) {
    channel.invokeMethod("onClose", arguments: nil)
  }

  deinit {
    // Release the handler block the messenger holds per channel name; the
    // [weak self] capture keeps the view collectable either way.
    channel.setMethodCallHandler(nil)
    ezoicNativeAd?.destroy()
  }
}
